### PR TITLE
 fix: Fix C++ decoder synchronization in PyTorch

### PR DESCRIPTION
**Pull request by dcampora**
This MR fixes C++ decoders in PyTorch:

* It introduces proper synchronization.
* It adheres to exposing methods decode_async / update_requests.
* Introduces send and recv tensor list.
* Introduces algorithm UpdateDecoderBuffers.
* Uses guided_decoder for all requests, acting as a slot manager.
* Follows the order: _forward_step, _decode_async, _update_request_states, _update_requests in py_executor.
* Overlap scheduling is now working as well.

### CODE REVIEW

**Comment by niukuo** [comment]
[PR_Github #605](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/605/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #605](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/605/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #512](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/512/) completed with status: 'FAILURE'

File: cpp/tensorrt_llm/pybind/runtime/bindings.cpp [hunk 1]
```diff
@@ -323,15 +324,9 @@ void initBindings(pybind11::module_& m)
...
-            [](CudaStreamPtr stream, std::vector<bool> const& active)
-            {
-                tr::CudaEvent eventStop{};
-                stream->record(eventStop);
-                return std::make_unique<tr::decoder_batch::DecoderFinishedEvent>(std::move(eventStop), active);
-            }))
-        .def("synchronize", [](tr::decoder_batch::DecoderFinishedEvent& self) { self.event.synchronize(); });
+    py::class_<tr::CudaEvent>(m, "CudaEvent")
+        .def(py::init<unsigned int>(), py::arg("flags") = cudaEventDisableTiming)
+        .def("synchronize", &tr::CudaEvent::synchronize);
```
Comment (Funatiq) [inline cpp/tensorrt_llm/pybind/runtime/bindings.cpp:344]: This conflicts with my changes [here](https://github.com/NVIDIA/TensorRT-LLM/pull/3195/files#diff-20f94577f09a2d7bdaa31bd2452065b0aae5f70c345e37f49feae4a8449d789bR339). I tried to keep the functionality of `DecoderFinishedEvent` but your change seems more straight forward.

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 2]
```diff
@@ -41,19 +36,26 @@ def decode_async(self, scheduled_requests: ScheduledRequests,
         raise NotImplementedError
 
     @abstractmethod
-    def update_requests(self, scheduled_requests: ScheduledRequests,
-                        new_tensors_host: Dict[str, torch.tensor],
-                        decoder_event: torch.cuda.Event):
+    def update_requests(self,
+                        scheduled_requests: ScheduledRequests,
+                        new_tensors_host: OrderedDict[str, torch.tensor],
+                        decoder_event: torch.cuda.Event or None = None):
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]:
```suggestion
                        decoder_event: Optional[torch.cuda.Event] = None):
```

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 3]
```diff
@@ -27,12 +27,7 @@
 class Decoder(ABC):
 
     @abstractmethod
-    def setup_decoder(self, scheduled_requests: ScheduledRequests,
-                      model_outputs):
-        raise NotImplementedError
-
-    @abstractmethod
-    def decode(self, scheduled_requests: ScheduledRequests, model_outputs):
+    def setup_decoder_step(self, requests):
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: It would be good to add a comment here explaining the difference between `scheduled_requests` and `requests`.

Should this function only be called from the decoder itself (in `decode_async`)? Then we should add a prefix `_`.

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 4]
```diff
@@ -70,25 +72,24 @@ class EarlyStopDecoder(Decoder):
...
             request.context_logits = model_outputs['logits'][idx]
+        return None, None, None
+
+    def update_requests(self,
+                        scheduled_requests,
+                        new_tensors_host,
+                        decoder_event: torch.cuda.Event or None = None):
+        assert (not scheduled_requests.generation_requests)
+        for request in scheduled_requests.context_requests:
+            request.state = LlmRequestState.GENERATION_COMPLETE
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: Doesn't that change the behavior in overlap? If we set the state in  `decode_async` then we won't schedule the requests for another iteration.

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 5]
```diff
@@ -523,38 +514,85 @@ def setup_decoder(self, scheduled_requests: ScheduledRequests,
...
+            dtype=torch.int,
+            device='cuda')
+        for idx, request in enumerate(
+                itertools.chain(scheduled_requests.context_requests,
+                                scheduled_requests.generation_requests)):
+            seq_slot = request.seq_slot
+            for beam in range(self.beam_width):
+                new_tokens_device_tensor[idx * self.beam_width + beam].copy_(
+                    self.algs.decoder.all_new_tokens[0][seq_slot][beam],
+                    non_blocking=True)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: Do we have to manually copy each individual token? Can't we do something like
```suggestion
            new_tokens_device_tensor[idx * self.beam_width:(idx + 1) * self.beam_width] = 
                self.algs.decoder.all_new_tokens[0][seq_slot][:self.beam_width]
```
or 
```suggestion
            new_tokens_device_tensor[idx] = self.algs.decoder.all_new_tokens[0][seq_slot]
```
or even
```
        seq_slots = [request.seq_slot for request in requests]
        new_tokens_device_tensor = self.algs.decoder.all_new_tokens[0][seq_slots]
```
(this needs some more work to make it non-blocking but it's much more pythonic and can be done in a single kernel. See previous_batch_indices_cuda in model_engine).

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 6]
```diff
@@ -523,38 +514,85 @@ def setup_decoder(self, scheduled_requests: ScheduledRequests,
...
-            self.decoder_event.record()
-
-    def update_requests(self, scheduled_requests: ScheduledRequests):
-        self.decoder_event.synchronize()
+        # NOTE: The following code prepares a new_tokens_device_tensor in accordance with the
+        #       current implementation of model_engine.
+        # TODO: When we support speculative decoding:
+        # new_tokens_device_tensor should be, for speculative decoding cases: [batch, 1 + draft_len], others: [batch]
+        new_tokens_device_tensor = torch.empty(
+            (self.batch_size * self.beam_width, ),
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: Why not make it 2-dim? That should make copying easier. I don't think we need the loop and can just gather the needed slots.

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py [hunk 7]
```diff
@@ -242,14 +242,15 @@ def create_py_executor(executor_config: ExecutorConfig,
...
+        if executor_config.guided_decoding_config is not None and spec_config is not None:
             raise ValueError(
-                "Guided decoding does not support with speculative decoding.")
+                "Guided decoding is not supported with speculative decoding.")
+
         resources[
             "guided_decoder_resource_manager"] = GuidedDecoderResourceManager(
-                executor_config.max_batch_size)
+                executor_config.max_batch_size +
+                pytorch_backend_config.enable_overlap_scheduler)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/py_executor_creator.py]: Why add 1 here?

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 3 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: This function is used in two places: one is decode_async, where context requests are used here, and the other is disaggregated serving, when the generation rank receives the generation requests, being invoked from py_executor.

I would argue the name is fine, however we should rename the parameter in trtGptModelIfb, which assumes context requests, which is imprecise.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 6 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: We can either adhere to the format of `new_tokens_device` in `model_engine.py`, or change it. If we adhere to it, this tensor must ultimately be one-dimensional. Or is your suggestion to make it 2D, copy, and then pass `.view(-1)`?

Either way, the better solution in the future would be to use the multi-dimensional object, whereby slots are not necessarily consecutive. From a first read of model_engine.py that does not seem straightforward to do.

File: cpp/tensorrt_llm/pybind/runtime/bindings.cpp (hunk 1 above)
Comment (dcampora) [reply cpp/tensorrt_llm/pybind/runtime/bindings.cpp:344]: I suggest to go with the simpler version here. I'll rebase.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 4 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: This line is changing the state in `update_requests`, not in `decode_async`, which should not impact overlap.

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py (hunk 7 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/py_executor_creator.py]: Needs more testing, but I would expect the sequence slot manager (here named GuidedDecoderResourceManager) to require one more slot when overlap scheduling is active.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 5 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: I agree with the above versions, I'll put a non-blocking version of one of them. Just to point out, right now beam search is not supported so copying tokens one by one or copying an entire slot is effectively doing the same thing.

Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: Done.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 4 above)
Comment (Funatiq) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: What I mean is the following:
```
_forward_step 0
_decode_async 0 // not setting complete 
_update_request_states 0
_forward_step 1 // requests will be scheduled again because they are not completed
_decode_async 1
_update_request_states 1
_update_requests 0 // setting complete here
```

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 6 above)
Comment (Funatiq) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: > Or is your suggestion to make it 2D, copy, and then pass .view(-1)?

Yes, that's what I meant. Btw there is `.flatten()` 😉

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 8]
```diff
@@ -25,14 +25,13 @@ def get_needed_resource_to_completion(self, request: LlmRequest) -> int:
...
         for llm_req in itertools.chain(scheduled_batch.context_requests,
                                        scheduled_batch.generation_requests):
-            if llm_req.guided_decoding_params is None:
-                continue
-            if llm_req.is_context_init_state and llm_req.context_current_position == llm_req.prepopulated_prompt_len:
-                self.slot_manager.add_slot(llm_req.request_id)
+            if llm_req.is_context_init_state and not llm_req.seq_slot or \
+                llm_req.is_disagg_generation_transmission_complete:
+                llm_req.seq_slot = self.slot_manager.add_slot(
+                    llm_req.request_id)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: I'm a bit worried that we are misusing `GuidedDecoderResourceManager` here. Guided decoding is a special use case. It would make more sense to have a dedicated `SeqSlotManager`.

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py (hunk 7 above)
Comment (Funatiq) [reply tensorrt_llm/_torch/pyexecutor/py_executor_creator.py]: Overlap scheduling should not increase the number of requests beyond the executor `max_batch_size` limit. The number of scheduled requests should never be larger than that.

**Comment by Funatiq** [comment]
Would it be possible to add some test for the new decoder path? It would make it easier to refactor the decoders without breaking the PyTorch implementation.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py (hunk 8 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: Do we need both? Why not rename `GuidedDecoderResourceManager` to `SeqSlotManager` instead?

Comment (Funatiq) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: I don't know if we need both. Renaming and moving it to another file would be fine with me.

Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: I just discussed this with @QiJune  and we agreed to rename it.

**Comment by dcampora** [comment]
> Would it be possible to add some test for the new decoder path? It would make it easier to refactor the decoders without breaking the PyTorch implementation.

Yes, I added a test in `test_overlap_scheduler.py` to cover running with the C++ decoders, temperature, stop words and overlap scheduling.

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 9]
```diff
@@ -64,7 +64,7 @@ def prepare_resources(self, scheduled_batch: ScheduledRequests):
         # allocate hidden state tensors
         for req in context_batch:
             if req.is_first_context_chunk():
-                self.slot_manager.add_slot(req.request_id)
+                req.seq_slot = self.slot_manager.add_slot(req.request_id)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/speculative/mtp.py]: There is a separate `SlotManager` in here. Should it really overwrite the `req.seq_slot` from the `SeqSlotManager` ?

**Review by Funatiq — COMMENTED** [review]
I noticed that `DecoderInputBuffers` is a member of the decoder here. In C++ runtime it is stored outside of the decoder and separate per batch so that it is not overwritten in overlapping iterations.

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 9 above)
Comment (dcampora) [reply tensorrt_llm/_torch/speculative/mtp.py]: Good catch, changed.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 6 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: This is done.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 4 above)
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: This comment refers to the implementation of `EarlyStopDecoder`. That decoder is not meant to support overlap scheduler (see its previous implementation [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/pyexecutor/decoder.py#L67-L91)). The only reason I moved the implementation to follow decode_async / update_requests is to unify the syntax (note that `decode` alone wouldn't make sense in the TRTLLMDecoder case, as that requires a step in between done by py_executor).

Comment (Funatiq) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: I see what you mean, but nothing prevents anyone to use it with overlap. Resolving for now.

**Review by Funatiq — APPROVED** [review]

File: 3rdparty/ucxx
Comment (Funatiq) [inline 3rdparty/ucxx:1]: Please remove the submodule update from this PR.

Comment (dcampora) [reply 3rdparty/ucxx:1]: Done.

**Review by byshiue — APPROVED** [review]

**Comment by dcampora** [comment]
Not sure what is going on. Tests that are run are passing. Some tests are not finished, but they don't seem to timeout.

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 10]
```diff
@@ -337,12 +328,13 @@ def _batch_decode(self, scheduled_requests: ScheduledRequests,
...
         decoder_event.record()
-        return DecoderState(
-            scheduled_requests=scheduled_requests,
-            logits=logits,
-            new_tensors_device={"new_tokens_device": new_tokens_device},
-            new_tensors_host={"new_tokens_host": new_tokens_host},
-            decoder_event=decoder_event)
+        return DecoderState(scheduled_requests=scheduled_requests,
+                            logits=logits,
+                            new_tensors_device=OrderedDict(
```
Comment (tongyuantongyu) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: `OrderedDict` does not seem necessary here. `dict` preserves insertion order starting from Python 3.6, and `OrderedDict` is only better at reordering elements which we don't need.

Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: Thanks, agreed. Fixed.

**Comment by dcampora** [comment]
/bot skip --comment "CI passed in last two CI runs."
